File: api/src/domain/repos/database.repo.ts
import type { CreateDatabaseInput, Database, UpdateDatabaseInput } from '../entities';

export interface IDatabaseRepo {
	create(database: CreateDatabaseInput): Promise<Database | null>;
	update(database: UpdateDatabaseInput): Promise<Database | null>;
	delete(id: string): Promise<void>;
	findById(id: string): Promise<Database | null>;
	findBySubdomainId(subdomainId: string): Promise<Database | null>;
}

File: api/internal/app/ports/database_repo.go
package ports

import (
	"api/internal/domain"
	"context"
)

type DatabaseRepo interface {
	FindByID(id string, context context.Context) (*domain.Database, error)
	// FindBySubdomainID returns the subdomain's database (1:1 today);
	// gorm.ErrRecordNotFound when the subdomain has none.
	FindBySubdomainID(subdomainID string, context context.Context) (*domain.Database, error)
	Create(db *domain.Database, context context.Context) error
	Update(id string, updates map[string]any, context context.Context) error
	Delete(db *domain.Database, context context.Context) error
}
